### PR TITLE
A first attempt at drawing pictures of WOM graphs

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

File: src/main/scala/wdltool/graph/WomGraphPrint.scala [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    val singleNode =
+      s"""
+         |subgraph $clusterName {
+         |  style=filled;
+         |  fillcolor=${graphFillColor(graphNode)};
+         |  ${portNodes.mkString(sep="\n  ")}
+         |  ${graphId(graphNode)} [shape=plaintext label=${graphName(graphNode)}]
+         |}
+        """.stripMargin
```
Comment (kshakir) [inline src/main/scala/wdltool/graph/WomGraphPrint.scala]: ToL

The last nine characters of this variable are `<newline><space><space><space><space><space><space><space><space>`.

This is a little bit neater. 

```scala
s"""|
    |subgraph $clusterName {
    |  style=filled;
    |  fillcolor=${graphFillColor(graphNode)};
    |  ${portNodes.mkString(sep="\n  ")}
    |  ${graphId(graphNode)} [shape=plaintext label=${graphName(graphNode)}]
    |}
    |""".stripMargin
```

File: src/main/scala/wdltool/graph/WomGraphPrint.scala [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+import cats.derived.monoid._
+import cats.derived.monoid.legacy._
+import cats.implicits._
+import wdl4s.cwl.{CommandLineTool, CwlCodecs, Workflow, Yaml}
+import wdl4s.wdl.{WdlNamespace, WdlNamespaceWithWorkflow}
+import wdl4s.wom.executable.Executable
+import wdl4s.wom.graph.GraphNodePort.{InputPort, OutputPort}
+import wdl4s.wom.graph.{GraphInputNode, GraphNodePort, GraphOutputNode}
+import wdltool.graph.GraphPrint.clusterCount
+//import wdl4s.wdl.{CallOutput, Declaration, If, Scatter, WdlCall, WdlGraphNode, _}
```
Comment (kshakir) [inline src/main/scala/wdltool/graph/WomGraphPrint.scala]: Remove this comment.

File: src/main/scala/wdltool/Main.scala [hunk 3]
```diff
@@ -140,13 +157,20 @@ object Main extends App {
...
+                       |
+                       |graph <WDL file>
                        |
                        |  Reads a WDL file against the grammar and prints out a
                        |  .dot of the DAG if it is valid, and a syntax error
                        |  otherwise.
-                       |  Use [--all] to show all graph nodes in the WDL spec,
-                       |  even the non-executable nodes.
+                       |
+                       | womgraph <WDL or CWL file> [ancillary files]
```
Comment (kshakir) [inline src/main/scala/wdltool/Main.scala]: There's an extra space before ` womgraph`.

**Comment by kshakir** [comment]
Minor issues. Otherwise 👍 

[Approved with PullApprove](https://pullapprove.com/broadinstitute/wdltool/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/wdltool)

File: src/main/scala/wdltool/graph/WomGraphPrint.scala [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        clt.womExecutable match {
+          case Valid(wom) => wom
+        }
+      case Right((wf: Workflow, m)) if m.isEmpty =>
+        wf.womExecutable(Map.empty) match {
+          case Valid(wom) => wom
+        }
+        // TODO: Don't throw away the error information! Might as well return it:
+      case other => throw new Exception(s"Can't read $filePath")
+    }
```
Comment (mcovarr) [inline src/main/scala/wdltool/graph/WomGraphPrint.scala]: ok so all these lines of boilerplate are somewhat suggestive of the need for a friendlier CWL API...

File: src/main/scala/wdltool/graph/WomGraphPrint.scala [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    */
+  private def defaultNodesAndLinks(graphNode: GraphNode): NodesAndLinks = {
+    val clusterName = "cluster_" + clusterCount.getAndIncrement()
+    val portNodes = (graphNode.outputPorts ++ graphNode.inputPorts) map { p =>
+      s"${graphId(p)} [shape=${graphShape(p)} label=${graphName(p)}];"
+    }
+    val links = for {
+      inputPort <- graphNode.inputPorts
+      upstreamPort = inputPort.upstream
+    } yield graphId(upstreamPort)+" -> "+graphId(inputPort)
```
Comment (mcovarr) [inline src/main/scala/wdltool/graph/WomGraphPrint.scala]: hit this line with some Command Option  L

File: src/main/scala/wdltool/graph/WomGraphPrint.scala [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        """.stripMargin
+
+    NodesAndLinks(Set(singleNode), links)
+  }
+
+
+  private def dotSafe(s: String) = s""""${s.replaceAllLiterally("\"", "\\\"")}""""
+
+  private def graphId(g: GraphNode): String = dotSafe("NODE"+g.hashCode().toString)
+  private def graphId(p: GraphNodePort): String = dotSafe("PORT"+p.hashCode().toString)
```
Comment (mcovarr) [inline src/main/scala/wdltool/graph/WomGraphPrint.scala]: ok maybe Command Option L the whole file

File: src/main/scala/wdltool/graph/WomGraphPrint.scala [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    case other =>
+      throw new Exception(s"Can't name $other")
+  })
+
+  private def graphShape(p: GraphNodePort) = p match {
+    case _: InputPort => "oval"
+    case _: OutputPort => "hexagon"
+  }
+
+  private def graphFillColor(g: GraphNode) = g match {
```
Comment (mcovarr) [inline src/main/scala/wdltool/graph/WomGraphPrint.scala]: ToL personally I would have `implicit class`ed all these methods for `GraphNode` and `GraphNodePort` to help readability

Comment (mcovarr) [reply src/main/scala/wdltool/graph/WomGraphPrint.scala]: e.g.

```
graphFillColor(graphNode)
```

becomes

```
graphNode.fillColor
```

**Review by mcovarr — APPROVED** [review]

**Comment by mcovarr** [comment]
oh 💩 this repo doesn't have pullapprove - thumb equivalency

:+1:

[Approved with PullApprove](https://pullapprove.com/broadinstitute/wdltool/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/wdltool)

File: src/main/scala/wdltool/graph/WomGraphPrint.scala (hunk 5 above)
Comment (cjllanwarne) [reply src/main/scala/wdltool/graph/WomGraphPrint.scala]: I read somewhere (and kinda agree) that when you have string concatenation it makes it way more readable (and easy to see where the final spaces will end up) to do this than to spread them out, eg
```
 val a = "Hi"
 val b = "Miguel"
 val c = a+" there "+b
// vs
 val d = a + " there " + b
```

OTOH, `s""` formatting is even better yet so I'll probably just change to that

Comment (mcovarr) [reply src/main/scala/wdltool/graph/WomGraphPrint.scala]: thumbsup to that last thing you said 😄

**Comment by cjllanwarne** [comment]
@mcovarr it appears that it is also never built by Travis since this is a floating wdl4s tag yet it's still green

File: src/main/scala/wdltool/graph/WomGraphPrint.scala (hunk 7 above)
Comment (geoffjentry) [reply src/main/scala/wdltool/graph/WomGraphPrint.scala]: I'm also a fan of such things

File: build.sbt [hunk 8]
```diff
@@ -8,7 +8,7 @@ organization := "org.broadinstitute"
 
 scalaVersion := "2.12.1"
 
-val wdl4sV = "0.14-7c693a3-SNAP"
+val wdl4sV = "0.14-44a952f-SNAP"
```
Comment (cjllanwarne) [inline build.sbt]: Fix to a published version before merging

**Pull request merged**